Neos.EventSourcedNeosAdjustments/Classes/EventSourcedRouting/Projection/DocumentUriPathFinder.php

```diff
@@ -0,0 +1,146 @@
+<?php
+declare(strict_types=1);
+
+namespace Neos\EventSourcedNeosAdjustments\EventSourcedRouting\Projection;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\DBALException;
+use Neos\ContentRepository\Domain\ContentStream\ContentStreamIdentifier;
+use Neos\ContentRepository\Domain\NodeAggregate\NodeAggregateIdentifier;
+use Neos\ContentRepository\Domain\NodeAggregate\NodeName;
+use Neos\EventSourcedNeosAdjustments\EventSourcedRouting\ValueObject\DocumentNodeInfo;
+use Neos\Flow\Annotations as Flow;
+use Neos\Neos\Controller\Exception\NodeNotFoundException;
+
+/**
+ * @Flow\Scope("singleton")
+ */
+final class DocumentUriPathFinder
+{
+    private Connection $dbal;
+
+    private ?ContentStreamIdentifier $liveContentStreamIdentifierRuntimeCache = null;
+
+    public function __construct(Connection $dbal)
+    {
+        $this->dbal = $dbal;
+    }
+
+    /**
+     * @param NodeName $siteNodeName
+     * @param string $uriPath
+     * @param string $dimensionSpacePointHash
+     * @return DocumentNodeInfo
+     * @throws NodeNotFoundException
+     */
+    public function getEnabledBySiteNodeNameUriPathAndDimensionSpacePointHash(NodeName $siteNodeName, string $uriPath, string $dimensionSpacePointHash): DocumentNodeInfo
+    {
+        return $this->fetchSingle('dimensionSpacePointHash = :dimensionSpacePointHash AND siteNodeName = :siteNodeName AND uriPath = :uriPath AND disabled = 0', compact('dimensionSpacePointHash', 'siteNodeName', 'uriPath'));
+    }
+
+    /**
+     * @param NodeAggregateIdentifier $nodeAggregateIdentifier
+     * @param string $dimensionSpacePointHash
+     * @return DocumentNodeInfo
+     * @throws NodeNotFoundException
+     */
+    public function getOneByIdAndDimensionSpacePointHash(NodeAggregateIdentifier $nodeAggregateIdentifier, string $dimensionSpacePointHash): DocumentNodeInfo
+    {
+        return $this->fetchSingle('nodeAggregateIdentifier = :nodeAggregateIdentifier AND dimensionSpacePointHash = :dimensionSpacePointHash', compact('nodeAggregateIdentifier', 'dimensionSpacePointHash'));
+    }
+
+    /**
+     * @param DocumentNodeInfo $nodeInfo
+     * @return DocumentNodeInfo
+     * @throws NodeNotFoundException
+     */
+    public function getParentNode(DocumentNodeInfo $nodeInfo): DocumentNodeInfo
+    {
+        return $this->getOneByIdAndDimensionSpacePointHash($nodeInfo->getParentNodeAggregateIdentifier(), $nodeInfo->getDimensionSpacePointHash());
+    }
+
+    /**
+     * @param NodeAggregateIdentifier $succeedingNodeAggregateIdentifier
+     * @param NodeAggregateIdentifier $parentNodeAggregateIdentifier
+     * @param string $dimensionSpacePointHash
+     * @return DocumentNodeInfo
+     * @throws NodeNotFoundException
+     */
+    public function getPrecedingNode(NodeAggregateIdentifier $succeedingNodeAggregateIdentifier, NodeAggregateIdentifier $parentNodeAggregateIdentifier, string $dimensionSpacePointHash): DocumentNodeInfo
```

Comment:
short comment to answer: "is preceding on the same hierarchy level or not"?

What is returned if there is no preceding node?

Reply:
Good point. Yes, it means: preceding on the same level and it will throw if no node matches this criteria. I added a doc comment explaining this